opm/output/data/GuideRateValue.hpp

```diff
@@ -0,0 +1,137 @@
+#ifndef OPM_OUTPUT_DATA_GUIDERATEVALUE_HPP
+#define OPM_OUTPUT_DATA_GUIDERATEVALUE_HPP
+
+#include <array>
+#include <bitset>
+#include <cstddef>
+#include <stdexcept>
+#include <string>
+
+namespace Opm { namespace data {
+
+    class GuideRateValue {
+    public:
+        enum class Item : std::size_t {
+            Oil, Gas, Water, ResV,
+
+            // -- Must be last enumerator --
+            NumItems,
+        };
+
+        void clear()
+        {
+            this->mask_.reset();
+            this->value_.fill(0.0);
+        }
+
+        constexpr bool has(const Item p) const
+        {
+            const auto i = this->index(p);
+
+            return (i < Size) && this->mask_[i];
+        }
+
+        bool operator==(const GuideRateValue& vec) const
+        {
+            return (this->mask_  == vec.mask_)
+                && (this->value_ == vec.value_);
+        }
+
+        double get(const Item p) const
+        {
+            if (! this->has(p)) {
+                throw std::invalid_argument {
+                    "Request for Unset Item Value for " + this->itemName(p)
+                };
+            }
+
+            return this->value_[ this->index(p) ];
+        }
+
+        GuideRateValue& set(const Item p, const double value)
+        {
+            const auto i = this->index(p);
+
+            if (i >= Size) {
+                throw std::invalid_argument {
+                    "Cannot Assign Item Value for Unsupported Item '"
+                    + this->itemName(p) + '\''
+                };
+            }
+
+            this->mask_.set(i);
+            this->value_[i] = value;
+
+            return *this;
+        }
+
+        GuideRateValue& operator+=(const GuideRateValue& rhs)
+        {
+            for (auto i = 0*Size; i < Size; ++i) {
+                if (rhs.mask_[i]) {
+                    this->mask_.set(i);
+                    this->value_[i] += rhs.value_[i];
+                }
+            }
+
+            return *this;
+        }
+
+        template <class MessageBufferType>
+        void write(MessageBufferType& buffer) const
+        {
+            auto maskrep = this->mask_.to_ullong();
+            buffer.write(maskrep);
+
+            for (const auto& x : this->value_) {
+                buffer.write(x);
+            }
+        }
+
+        template <class MessageBufferType>
+        void read(MessageBufferType& buffer)
+        {
+            this->clear();
+
+            {
+                auto mask = 0ull;
+                buffer.read(mask);
+
+                this->mask_ = std::bitset<Size>(mask);
+            }
+
+            for (auto& x : this->value_) {
+                buffer.read(x);
+            }
+        }
+
+    private:
+        enum { Size = static_cast<std::size_t>(Item::NumItems) };
```

Comment:
I would prefer if the `Size` was construction was simplified to e.g.
```C++
constexpr std::size_t size() const {
     return static_cast<std::size_t>(Item::Resv) + 1;
}
```
or something similar - the current construction using an enum to denote size is not very intuitive.

Reply:
We must have a compile-time constant that does not require an object since we're using it (`Size`) as a template parameter for `std::bitset<>` and `std::array<>`.  Creating a private `enum` is the easiest way to generate one of those.  Playing games with `static` doesn't work either, as exemplified by the class `K` [below](https://godbolt.org/z/3zxe6d).  As for using a separate enumerator (`NumItems`) to infer the number of enumerators in `Item`, well that's a matter of taste I suppose.  It is nevertheless a standard technique to reduce maintenance cost until reflection allows us to iterate over the enumerators at compile time.

---

```C++
#include <bitset>
#include <cstddef>

class K
{
public:
    enum I : std::size_t { a, b, c };

    static constexpr std::size_t size()
    {
        return c + 1;
    }

    using bs = std::bitset<size()>;

private:
    bs bs_;
};
```

Reply:
?

Reply:
> ?

Did you follow the link to Compiler Explorer?  Then you'd see build failures along the lines of
```
./t.cpp:14:32: error: ‘static constexpr std::size_t K::size()’ called in a constant expression before its definition is complete
   14 |     using bs = std::bitset<size()>;
      |                            ~~~~^~
./t.cpp:14:32: note: in template argument for type ‘long unsigned int’
./t.cpp:17:5: error: ‘bs’ does not name a type; did you mean ‘abs’?
   17 |     bs bs_;
      |     ^~
      |     abs
```